File: solutions/Reverse-Linked-List/Reverse-Linked-List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function (head) {
  let newHead = null;
  while (head !== null) {
    const next = head.next;
    head.next = newHead;
    newHead = head;
    head = next;
  }
  return newHead;
};

/*
 * Recursively
 */

/*
 var reverseListHelper = function(curHead, prevHead) {
 if(curHead === null) {
 return prevHead;
 }
 const next = curHead.next;
 curHead.next = prevHead;
 return reverseListHelper(next, curHead);
 };

 var reverseList = function(head) {
 return reverseListHelper(head, null);
 };

 */

File: solutions/Reverse-Linked-List/Reverse-Linked-List.cpp
// 2019-07-28
// O(n)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* newHead = NULL;
        ListNode* p = head;
        while(p) 
        {
            ListNode* tmp = p->next;
            p->next = newHead;
            newHead = p;
            p = tmp;
        }
        return newHead;
    }
};